src/job_cache/types.cpp

```diff
@@ -107,62 +107,112 @@ JAST JobOutputInfo::to_json() const {
   return json;
 }
 
-MatchingJob::MatchingJob(const JAST &json) {
+MatchingJob::MatchingJob(const std::string &client_cwd, const JAST &json) {
   output_info = JobOutputInfo(json.get("output_info"));
 
   for (const auto &output_file_json : json.get("output_files").children) {
-    output_files.push_back(CachedOutputFile(output_file_json.second));
+    CachedOutputFile output_file(output_file_json.second);
+    // Canoncolize matching jobs to use client-relative paths.
+    if (wcl::is_absolute(output_file.path)) {
+      output_file.path = wcl::relative_to(client_cwd, output_file.path);
+    }
+    output_files.push_back(output_file);
   }
 
   for (const auto &output_dir_json : json.get("output_dirs").children) {
-    output_dirs.push_back(CachedOutputDir(output_dir_json.second));
+    CachedOutputDir output_dir(output_dir_json.second);
+    // Canoncolize matching jobs to use client-relative paths.
+    if (wcl::is_absolute(output_dir.path)) {
+      output_dir.path = wcl::relative_to(client_cwd, output_dir.path);
+    }
+    output_dirs.push_back(output_dir);
   }
 
   for (const auto &output_symlink_json : json.get("output_symlinks").children) {
-    output_symlinks.push_back(CachedOutputSymlink(output_symlink_json.second));
+    CachedOutputSymlink output_sym(output_symlink_json.second);
+    // Canoncolize matching jobs to use client-relative paths.
+    if (wcl::is_absolute(output_sym.path)) {
+      output_sym.path = wcl::relative_to(client_cwd, output_sym.path);
+    }
+    output_symlinks.push_back(output_sym);
   }
 
   for (const auto &input_file_json : json.get("input_files").children) {
-    input_files.push_back(input_file_json.second.value);
+    // Canoncolize matching jobs to use client-relative paths.
+    std::string input_path = input_file_json.second.value;
+    if (wcl::is_absolute(input_path)) {
+      input_path = wcl::relative_to(client_cwd, input_path);
+    }
+    input_files.push_back(input_path);
   }
 
   for (const auto &input_dir_json : json.get("input_dirs").children) {
-    input_dirs.push_back(input_dir_json.second.value);
+    // Canoncolize matching jobs to use client-relative paths.
```

Comment:
sp: Canonicalize

Reply:
lol  this is in here a bunch, i'll stop making new comments but it should be fixed everywhere